### PR TITLE
DataVolume functional test suite

### PR DESCRIPTION
**Pull request by davidvossel**
This PR is adding initial functional tests for DataVolumes into cdi

I've added hostpath PVs and an http import server to help with testing. This will help the importer and clone tests as well.

### CODE REVIEW

**Comment by davidvossel** [comment]
This is no longer a WIP. It should be ready for review now.

File: cluster/sync.sh [hunk 1]
```diff
@@ -28,3 +28,5 @@ for i in $(seq 1 ${KUBEVIRT_NUM_NODES}); do
 done
 
 ./cluster/kubectl.sh apply -f ./manifests/generated/cdi-controller.yaml
+# this http server is used by the functional tests
+./cluster/kubectl.sh apply -f ./manifests/generated/disks-images-provider.yaml
```
Comment (jeffvance) [inline cluster/sync.sh]: how does this fit in with the server provided by pr #315 ?

**Comment by jeffvance** [comment]
@copejon need review from you due to your pr #315

File: manifests/templates/disks-images-provider.yaml.in [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    kubevirt.io: ""
+    os: "datavolume1"
+spec:
+  capacity:
+    storage: 1Gi
+  accessModes:
+    - ReadWriteOnce
+  hostPath:
+    path: /tmp/hostImages/datavolume1
+  persistentVolumeReclaimPolicy: Recycle
```
Comment (jeffvance) [inline manifests/templates/disks-images-provider.yaml.in]: Do you really want recycle? Also, AFAIK recycle will be deprecated (though that's been a long road).

File: cluster/sync.sh (hunk 1 above)
Comment (davidvossel) [reply cluster/sync.sh]: This is a separate approach. I use a container with nginx in it to serve the static iso files rather that writing code to do that. 

It's hard for me to justify using maintaining code for an http server when we can accomplish the same thing with little effort using nginx. Is there any benefit to the http server in #315 over using nginx? Maybe I'm missing something?

File: manifests/templates/disks-images-provider.yaml.in [hunk 3]
```diff
@@ -0,0 +1,92 @@
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  name: host-path-disk-datavolume1
+  labels:
+    kubevirt.io: ""
+    os: "datavolume1"
```
Comment (jeffvance) [inline manifests/templates/disks-images-provider.yaml.in]: what is the `os` label for?

File: tests/utils.go [hunk 4]
```diff
@@ -5,22 +5,35 @@ import (
...
+	clientset "kubevirt.io/containerized-data-importer/pkg/client/clientset/versioned"
+	"kubevirt.io/containerized-data-importer/pkg/util"
 )
 
 var KubectlPath = ""
 var OcPath = ""
 var CDIInstallNamespace = "kube-system"
 
+const (
+	AlpineUrl         = "http://cdi-http-import-server.kube-system/images/alpine.iso"
```
Comment (jeffvance) [inline tests/utils.go]: all of our existing tests using tinyCore.iso. I like the idea of including additional images. Should we add alpine.iso to _test/images_?

File: tests/utils.go [hunk 5]
```diff
@@ -42,6 +55,16 @@ func CDIFailHandler(message string, callerSkip ...int) {
...
 }
 
+func GetCDIClient() *clientset.Clientset {
+
+	cfg, err := clientcmd.BuildConfigFromFlags(master, kubeconfig)
+	PanicOnError(err)
+	cdiClient, err := clientset.NewForConfig(cfg)
+	PanicOnError(err)
+
+	return cdiClient
```
Comment (jeffvance) [inline tests/utils.go]: Similar to k8s, I'd like to see an e2e framework struct that contains ns, client(s), provider, maybe some storage labels, etc so that these values are available to all tests. See https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/framework.go#L136 and `BeforeEach` just below.  Unless, there's a good kubevirt reason not to...

File: tests/utils.go [hunk 6]
```diff
@@ -100,3 +123,27 @@ func GetKubeClientFromRESTConfig(config *rest.Config) (*kubernetes.Clientset, er
 
 	return coreClient, nil
 }
+
+func GenerateNamespace() string {
```
Comment (jeffvance) [inline tests/utils.go]: I like this approach but we need to reconcile it with recently merged pr #314 https://github.com/kubevirt/containerized-data-importer/blob/master/tests/utils.go#L22 where the test hard-codes the ns

File: manifests/templates/disks-images-provider.yaml.in (hunk 2 above)
Comment (davidvossel) [reply manifests/templates/disks-images-provider.yaml.in]: the PVs are static in this setup, so we'll need recycling to use them with multiple PVCs throughout the test suite.

File: manifests/templates/disks-images-provider.yaml.in (hunk 3 above)
Comment (davidvossel) [reply manifests/templates/disks-images-provider.yaml.in]: that's now we match a PVC with the PV. we're using a selector called "os".  I pulled this logic from KubeVirt where "os" made more sense in their context than it does in CDI.

File: tests/utils.go (hunk 4 above)
Comment (davidvossel) [reply tests/utils.go]: Oh, i missed that tinyCore.iso already exists in our repo.  The http importer service container should be serving the images in test/images/ instead of downloading alpine. 

so no, we don't need alpine here I don't think. the isos already in the src tree should be fine

File: tests/utils.go (hunk 6 above)
Comment (davidvossel) [reply tests/utils.go]: Its fine to use hard coded namespaces.  We'll need more than one though in order to test cloning across multiple namespaces.

Comment (awels) [reply tests/utils.go]: That is the install namespace, where when we don't specify a different name space in the manifest generation the cdi stuff is installed. It is setup as an argument to the test runner so we can override it without recompiling the code.

File: cluster/sync.sh (hunk 1 above)
Comment (awels) [reply cluster/sync.sh]: @davidvossel I think the difference here, is that #315 uses one source file, and it generates all combinations of file formats we need, eg: tar/tar.gz/raw/etc for testing. We could do the same thing with nginx if we have an instance for each combination that it can serve. The difference would be some work to create all the files, ahead of time vs work to maintain the file server to support all the different format.

File: manifests/templates/disks-images-provider.yaml.in (hunk 3 above)
Comment (awels) [reply manifests/templates/disks-images-provider.yaml.in]: Do we really need to create PVs any more? We should have 10 local volume PVs available from the kubevirtci image at this point that we can leverage.

File: manifests/templates/disks-images-provider.yaml.in (hunk 2 above)
Comment (awels) [reply manifests/templates/disks-images-provider.yaml.in]: The local volume PVs from kubevirtci are also set to recycle.

File: manifests/templates/disks-images-provider.yaml.in (hunk 3 above)
Comment (davidvossel) [reply manifests/templates/disks-images-provider.yaml.in]: @awels some dev documentation would be helpful there. Something that just outlines what providers contain local storage support now, and an example spec for how to construct a PVC to ultilize local storage provider.

File: tests/utils.go [hunk 7]
```diff
@@ -42,6 +55,16 @@ func CDIFailHandler(message string, callerSkip ...int) {
 	Fail(message, callerSkip...)
 }
 
+func GetCDIClient() *clientset.Clientset {
+
+	cfg, err := clientcmd.BuildConfigFromFlags(master, kubeconfig)
```
Comment (awels) [inline tests/utils.go]: You should be able to call tests.GetKubeClient() to get the same thing.

File: manifests/templates/disks-images-provider.yaml.in (hunk 3 above)
Comment (awels) [reply manifests/templates/disks-images-provider.yaml.in]: @davidvossel https://github.com/kubevirt/kubevirtci/pull/35 for the documentation.

File: tests/utils.go [hunk 8]
```diff
@@ -42,6 +55,16 @@ func CDIFailHandler(message string, callerSkip ...int) {
 	Fail(message, callerSkip...)
 }
 
+func GetCDIClient() *clientset.Clientset {
```
Comment (awels) [inline tests/utils.go]: How is this result any different from GetKubeClient below?

Comment (davidvossel) [reply tests/utils.go]: the cdi client is different from the kubernetes client.

cdi client knows about cdi objects (like DataVolumes) and the kubernetes client only knows about k8s core objects (like pods, pvcs, etc...) They are separate objects.

Comment (awels) [reply tests/utils.go]: Okay maybe I am missing something, but both use the same master and kubeconfig parameters. And they call the same functions to create the ClientSet, so they should be identical when we are running the test. Well except that the GetKubeClient adds some api stuff.

File: cluster/sync.sh (hunk 1 above)
Comment (copejon) [reply cluster/sync.sh]: @davidvossel Fair point - I don't know why I hadn't considered using nginx.  

As for the generation of files, that code exists as a separate binary to be run inside an init container.  It uses the same FormatTestData process from our unit tests, which keeps the addition of new code minimal.  The way I've implemented it, there are 3 containers - init, http, and s3.  They share access to the generated files through a shared emptyDir volume, so there's no duplication required.

Do you think the init process would be useful in the context of this PR?  If so, i can scrap the http server in favor of your implementation here.

File: manifests/templates/disks-images-provider.yaml.in (hunk 3 above)
Comment (copejon) [reply manifests/templates/disks-images-provider.yaml.in]: @awels 
> We should have 10 local volume PVs available from the kubevirtci image at this point that we can leverage.

Is the implication here that functional tests will be closely coupled w/ the kubevirtCI env by expecting PVs to exist already?

File: tests/utils.go (hunk 6 above)
Comment (copejon) [reply tests/utils.go]: @davidvossel 
> Its fine to use hard coded namespaces. We'll need more than one though in order to test cloning across multiple namespaces.

Is this still true when using ginkgo in parallel?

File: manifests/templates/disks-images-provider.yaml.in (hunk 3 above)
Comment (awels) [reply manifests/templates/disks-images-provider.yaml.in]: Not any more than us expecting a kubernetes environment to exist, I don't think its unreasonable for our tests to expect some kind of PV provider exists in the same environment, as these are CDI related tests.

Comment (davidvossel) [reply manifests/templates/disks-images-provider.yaml.in]: @awels awesome, thanks! I updated this PR to use local storage now. works great, this is super useful. I'll be using it in KubeVirt soon as well

File: cluster/sync.sh (hunk 1 above)
Comment (davidvossel) [reply cluster/sync.sh]: @copejon ah, yes i understand better what's going on now.

Yeah, I'd suggest pulling in the nginx server approach i took here into your work. Once your PR is merged, I'll scrap the http server in this PR and leverage what you've provided.  Sound good?

File: manifests/templates/disks-images-provider.yaml.in (hunk 3 above)
Comment (jeffvance) [reply manifests/templates/disks-images-provider.yaml.in]: @awels wrote:
>  I don't think its unreasonable for our tests to expect some kind of PV provider exists in the same environment

My background for e2e is in k8s. In that env the test ensures that the storage class, or pv or whatever storage is needed is explicitly created in the test setup (BeforeEach is common). If I understand what's going on here correctly, the test image is providing the pvs. If true that is quite different from k8s. Again, if true, what is the motivation for doing storage this way?

File: tests/utils.go (hunk 6 above)
Comment (jeffvance) [reply tests/utils.go]: @davidvossel wrote:
> ts fine to use hard coded namespaces.

How will that scale and work when tests are run in parallel?

File: manifests/templates/disks-images-provider.yaml.in (hunk 3 above)
Comment (awels) [reply manifests/templates/disks-images-provider.yaml.in]: For k8s e2e it would make sense for the test to create storage/whatever resources is needed for the test. As the test is exercising the framework that is using that particular resource. 

In our case it is different, we are testing something that is running on top of k8s, and expects a working k8s environment. Typically in a working k8s environment there is some kind of storage.

The motivation for doing it like this? We found that everything that was using the kubevirtci image needed some kind of storage, and everyone was doing something different (most of the time a hostPath based PV). So instead of having everyone reinvent the wheel each time, we implemented some basic local volume based storage in the image. Now everyone has some default storage available.

File: hack/build/config.sh [hunk 9]
```diff
@@ -35,7 +37,7 @@ function allPkgs {
     echo "$ret"
 }
 
-KUBERNETES_IMAGE="k8s-1.10.4@sha256:09ac918cc16f13a5d0af51d4c98e3e25cbf4f97b7b32fe18ec61b32f04ca1009"
+KUBERNETES_IMAGE="k8s-1.10.4@sha256:486064eddea289b17e150e6600fefc89dab9164d5cba07153c02888a35fed4f1"
```
Comment (copejon) [inline hack/build/config.sh]: Incrementing the k8s version should probably be isolated in its own pr.

File: tests/utils.go (hunk 6 above)
Comment (davidvossel) [reply tests/utils.go]: we'd have to randomize if we start running in parallel.

I'm not sure we'll want to run in parallel though. KubeVirt doesn't.

If we test in parallel, it impacts the timeouts associated with our tests. It becomes difficult to distinguish between an actual performance degrading timing issue and one that is just caused by multiple intensive tests running at the same time.  Also, parallel is more difficult to debug. It's the difference from having a controller log that's isolated to a single test and one that's got lots of unrelated things going on within it.

These functional tests are difficult to keep stable at times. Each k8s and os provider behaves slightly differently. Anything we can do to keep the execution of these tests as simple as possible will work in our favor.

File: tests/utils.go (hunk 8 above)
Comment (davidvossel) [reply tests/utils.go]: >  And they call the same functions to create the ClientSet

cdiClient is calling a package that's in the cdi src tree. Take a closer look at the clientset import at the top of the file.

File: tests/datavolume_test.go [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+	Describe("Verify DataVolume", func() {
+		It("create and import should succeed", func() {
+			newDataVolume := generateNewDataVolume("1Gi", name, tests.AlpineUrl, tests.HostPathSelector1)
+
+			By("creating new datavolume")
+			_, err := cdiClient.CdiV1alpha1().DataVolumes(namespace).Create(newDataVolume)
+			Expect(err).To(BeNil())
+
+			seconds := 30
```
[the remark this answers is not in the document]
Comment (davidvossel) [reply tests/datavolume_test.go]: The datavolume controller doesn't time out the import job. This is just an arbitrary timeout that's relatively relaxed.

Given the size of the image being imported and the size of the volume, we'd expect something wrong to have occurred if it took longer that 30 seconds to complete.

File: hack/build/config.sh (hunk 9 above)
Comment (davidvossel) [reply hack/build/config.sh]: yep, i submitted a separate pr for this. This will get resolve the next time i rebase this pr.

File: tests/utils.go (hunk 8 above)
Comment (awels) [reply tests/utils.go]: I understand, I totally missed that little difference. Thanks, the Framework patch I will push shortly will have both kubeclient and cdiclient available by default.

**Comment by copejon** [comment]
@davidvossel besides a rebase, is there anything in here that still needs to be addressed?

**Comment by davidvossel** [comment]
>  besides a rebase, is there anything in here that still needs to be addressed?

@copejon nope. the rebase is the big one. I've been waiting on all the refactoring work to land in master first. That will let me remove quite of a bit of what's already in this pr.

**Comment by jeffvance** [comment]
Hi David,

Thanks for your patience! I think the two remaining prs to get merged
before your massive rebase are 352 and 353, with 353 being more critical,
imo.

Jeff

**Comment by awels** [comment]
@davidvossel Looking at this patch, the framework should have all the utility functions you need. Only potential problem I see, is if the http server is not visible in all namespaces.

**Comment by davidvossel** [comment]
>  is if the http server is not visible in all namespaces

I'm re-writing this right now and haven't had any issues yet with the http server. Seems to be accessible throughout the cluster if I reference the Service DNS entry. 

example
```
wget http://cdi-file-host.kube-system/tinyCore.iso
```

**Comment by davidvossel** [comment]
@awels This should be good now for the initial dataVolume import functional tests. Can you take a look and see if I used all the test framework stuff right?

File: tests/datavolume_test.go [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+	"kubevirt.io/containerized-data-importer/tests/framework"
+	"kubevirt.io/containerized-data-importer/tests/utils"
+
+	cdiv1 "kubevirt.io/containerized-data-importer/pkg/apis/datavolumecontroller/v1alpha1"
+)
+
+var _ = Describe("DataVolume tests", func() {
+	flag.Parse()
+
+	f, err := framework.NewFramework("dv-func-test", &framework.Config{})
```
Comment (awels) [inline tests/datavolume_test.go]: We changed the pointer to a regular Config, so don't need the & anymore.

File: tests/datavolume_test.go [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"kubevirt.io/containerized-data-importer/tests/framework"
+	"kubevirt.io/containerized-data-importer/tests/utils"
+
+	cdiv1 "kubevirt.io/containerized-data-importer/pkg/apis/datavolumecontroller/v1alpha1"
+)
+
+var _ = Describe("DataVolume tests", func() {
+	flag.Parse()
```
Comment (awels) [inline tests/datavolume_test.go]: You shouldn't need the flag.Parse anymore, the framework calls it.

**Comment by awels** [comment]
@davidvossel looks good, two really minor comments

File: tests/framework/datavolume.go [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+	"kubevirt.io/containerized-data-importer/tests/utils"
+
+	cdiv1 "kubevirt.io/containerized-data-importer/pkg/apis/datavolumecontroller/v1alpha1"
+)
+
+func (f *Framework) CreateDataVolumeFromDefinition(dataVolume *cdiv1.DataVolume) (*cdiv1.DataVolume, error) {
+	return utils.CreateDataVolumeFromDefinition(f.CdiClient, f.Namespace.Name, dataVolume)
+}
+
+func (f *Framework) DeleteDataVolume(dataVolume *cdiv1.DataVolume) error {
```
Comment (jeffvance) [inline tests/framework/datavolume.go]: As a small added value this could be changed to `CreateDataVolumeFromDefinitionOrDie` and not return _error_,  test the err return from `utils.CreateDataVolumeFromDefinition`. 
Are you and @awels in agreement on what separates code in _framework_ vs _utils_? So far, to me, it looks like _framework_ defines and instantiates a *Framework and defines convenience metrhods that call the same func in _utils_ passing some fields from *Framework.

File: tests/datavolume_test.go (hunk 12 above)
Comment (jeffvance) [reply tests/datavolume_test.go]: Right. And all flag values are defined in the Framework struct.

File: tests/datavolume_test.go [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+	f, err := framework.NewFramework("dv-func-test", &framework.Config{})
+	if err != nil {
+		Fail("Unable to create framework struct")
+	}
+
+	Describe("Verify DataVolume", func() {
+		It("create and import should succeed", func() {
+			dataVolume := utils.NewDataVolumeWithHttpImport("test-dv1", "1Gi", utils.TinyCoreIsoURL)
+			By("creating new datavolume")
```
Comment (jeffvance) [inline tests/datavolume_test.go]: consider adding the name of the dv in the text, here and below?

File: tests/datavolume_test.go [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+			f.WaitForDataVolumePhase(cdiv1.Succeeded, dataVolume.Name)
+
+			// verify PVC was created
+			By("verifying pvc was created")
+			_, err = f.K8sClient.CoreV1().PersistentVolumeClaims(dataVolume.Namespace).Get(dataVolume.Name, metav1.GetOptions{})
+			Expect(err).To(BeNil())
+
+			err = f.DeleteDataVolume(dataVolume)
+			Expect(err).To(BeNil())
+		}, 45)
```
Comment (jeffvance) [inline tests/datavolume_test.go]: confused on timeout value here and below since there is no Done channel defined?

File: tests/datavolume_test.go [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+	})
+})
```
Comment (jeffvance) [inline tests/datavolume_test.go]: The 2 `It`s are very similar and could be defined in a simple test table, `[]struct{...}`. See #376 as an example.

File: tests/utils/datavolume.go [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+						k8sv1.ResourceName(k8sv1.ResourceStorage): resource.MustParse(size),
+					},
+				},
+			},
+		},
+	}
+}
+
+// Wait for the DataVolume to be in a particular phase (Pending, Bound, or Lost)
+func WaitForDataVolumePhase(clientSet *cdiclientset.Clientset, namespace string, phase cdiv1.DataVolumePhase, dataVolumeName string) error {
```
Comment (jeffvance) [inline tests/utils/datavolume.go]: This code this was copied from had a bug, now fixed in pr #378. You want something like this here:
```
func WaitForDataVolumePhase(clientSet *cdiclientset.Clientset, namespace string, phase cdiv1.DataVolumePhase, dataVolumeName string) error {
	wait.PollImmediate(dataVolumePollInterval, DataVolumePhaseTime, func() (bool, error) {
		dataVolume, err := clientSet.CdiV1alpha1().DataVolumes(namespace).Get(dataVolumeName, metav1.GetOptions{})
		if err != nil || dataVolume.Status.Phase != phase {
			return false, err
		}
		return true, nil
	})
	if err != nil {
            return fmt.Errorf("DataVolume %s not in phase %s within %v", dataVolumeName, phase, DataVolumePhaseTime)
        }
        return nil
}
```

File: tests/datavolume_test.go (hunk 16 above)
Comment (davidvossel) [reply tests/datavolume_test.go]: ginko provides a table extension. It's used in kubevirt. https://github.com/kubevirt/kubevirt/blob/master/tests/vnc_test.go#L123

Comment (jeffvance) [reply tests/datavolume_test.go]: Thanks David! I didn't know about native ginkgo tables.

File: tests/framework/datavolume.go (hunk 13 above)
Comment (davidvossel) [reply tests/framework/datavolume.go]: > Are you and @awels in agreement on what separates code in framework vs utils? 

@awels I'd be in favor of only adding functions to the framework package if the functions do something beyond simply calling something in the utils package directly.

**Comment by jeffvance** [comment]
@davidvossel needs rebase. Can you pls take a look. If merge is simple I can do it via github.

**Comment by davidvossel** [comment]
@awels @jeffvance feedback should be addressed now.  I took out the datavolume.go file in the framework package and justed used the utils package directly.

**Comment by awels** [comment]
lgtm

**Comment by jeffvance** [comment]
lgtm. Will merge after ci finishes.

**Comment by awels** [comment]
Random thought, I noticed Jon is checking to make sure the pods for the http server are up and running before running his tests. Are we going to be sure they will be up and running in time in a CI system?

**Comment by davidvossel** [comment]
> Are we going to be sure they will be up and running in time in a CI system?

whatever automation is performing the 'make cluster-up && make cluster-sync' needs to wait on running the functional tests until all pods related to CDI are both running and ready.

**Comment by awels** [comment]
Yeah, that is a good point, since I am working on that right now, let me make sure I put in a check to make sure they are up.
